### PR TITLE
Improve storybook build performance

### PR DESCRIPTION
**Pull request by vankop**
**What this PR does / why we need it**:

I wanted to improve Graph component in our fork regarding to https://github.com/grafana/grafana/issues/26523 and found out that developing using current storybook setup veeeery slow.. So I decided to improve build a little bit:

- upgrade loaders versions (`expose-loader` improve perf a lot)
- use react-docgen-typescript-loader only for production

**Which issue(s) this PR fixes**:

no issue

Fixes #

**Special notes for your reviewer**:

|       | cold start | incremental build |
|----|-----------|------------------|
| `master` https://github.com/grafana/grafana/commit/398bc045fc0094d9a1482f040d7ef334c6a8034b | 1.73 min for manager and 3.1 min for preview | 5-12s depending on changes |
|  PR           | 56s , 1.38min | 5-11.5s depending on changes |
| PR + transpileOnly | 25s, 45s | <4.5s |

This PR improves mostly only cold start. To improve incremental build I strongly recommend to move out of `ts-loader` and use babel instead or solve `ts-loader` issues. Also shipping es6+ in dev allow to use `options.devtool = false`

Profile of HMR update below
Снимок экрана 2020-08-08 в 1 00 15

### CODE REVIEW

File: packages/grafana-ui/src/components/Select/utils.ts [hunk 1]
```diff
@@ -20,19 +20,29 @@ export const cleanValue = (value: any, options: SelectOptions): SelectableValue[
...
 export const findSelectedValue = (value: string | number, options: SelectOptions): SelectableValue | null => {
-  for (const option of options) {
-    if ('options' in option) {
-      let found = findSelectedValue(value, option.options);
-      if (found) {
-        return found;
-      }
-    } else if ('value' in option && option.value === value) {
-      return option;
+  if (Array.isArray(options)) {
```
Comment (vankop) [inline packages/grafana-ui/src/components/Select/utils.ts:40]: found out this when was playing increasing `tsconfig.json` target to es2017

File: public/app/plugins/panel/gettingstarted/types.ts [hunk 2]
```diff
@@ -24,6 +24,6 @@ export interface SetupStep {
   subheading: string;
   title: string;
   info: string;
-  cards: (Card | TutorialCardType)[];
+  cards: Array<Card | TutorialCardType>;
```
Comment (vankop) [inline public/app/plugins/panel/gettingstarted/types.ts:27]: yarn lint

**Comment by aknuds1** [comment]
The front-end tests [fail](https://drone.grafana.net/grafana/grafana/1070/1/12).

**Comment by vankop** [comment]
> The front-end tests fail.

hm.. whatever, I have reverted utils changes

**Comment by aknuds1** [comment]
The StoryBook build looks really fast with your changes, but I'll leave it to the front-end team to look them over.

**Comment by torkelo** [comment]
Awesome work!  Big thanks from the frontend platform squad! :heart: we will look at this asap in the coming 2-3 days

File: packages/grafana-ui/.storybook/webpack.config.js [hunk 3]
```diff
@@ -7,30 +7,39 @@ module.exports = ({ config, mode }) => {
...
-          loader: require.resolve('react-docgen-typescript-loader'),
-          options: {
-            tsconfigPath: path.resolve(__dirname, 'tsconfig.json'),
-            // https://github.com/styleguidist/react-docgen-typescript#parseroptions
-            // @ts-ignore
-            propFilter: prop => {
-              if (prop.parent) {
-                return !prop.parent.fileName.includes('node_modules/@types/react/');
-              }
+      use: isProductionBuild
```
Comment (dprokop) [inline packages/grafana-ui/.storybook/webpack.config.js:10]: Good call, but this will result in the props table not being rendered in the dev mode's preview. I suggest maybe use env variables instead of the production assumption only (https://storybook.js.org/docs/configurations/env-vars/) and introduce another storybook script in the grafana-ui package.json that would allow opting in to render the docs tables too? like `yarn storybook:docs`?

File: packages/grafana-ui/.storybook/webpack.config.js [hunk 4]
```diff
@@ -65,16 +74,10 @@ module.exports = ({ config, mode }) => {
...
-      },
-      {
-        loader: 'expose-loader',
-        query: '$',
-      },
-    ],
+    loader: 'expose-loader',
+    options: {
+      exposes: ['jQuery', '$'],
+    },
```
Comment (dprokop) [inline packages/grafana-ui/.storybook/webpack.config.js:80]: 👍

File: packages/grafana-ui/.storybook/webpack.config.js [hunk 5]
```diff
@@ -89,7 +92,6 @@ module.exports = ({ config, mode }) => {
           test: /[\\/]node_modules[\\/].*[jt]sx?$/,
           chunks: 'initial',
           priority: -10,
-          reuseExistingChunk: true,
```
Comment (dprokop) [inline packages/grafana-ui/.storybook/webpack.config.js:92]: Why is this removed?

File: packages/grafana-ui/tsconfig.json [hunk 6]
```diff
@@ -1,7 +1,6 @@
 {
   "compilerOptions": {
     "baseUrl": "node_modules/@types",
-    "declarationDir": "dist",
```
Comment (dprokop) [inline packages/grafana-ui/tsconfig.json:4]: This will make the dist package not include declaration files, so it can not be removed unless we programmatically move all *.d.ts files when the package is bundled.

**Review by dprokop — CHANGES_REQUESTED** [review]
Thanks for the contribution @vankop. It's indeed faster, but it also introduces a couple of regressions that I've mentioned in my comments.

File: packages/grafana-ui/.storybook/webpack.config.js (hunk 3 above)
Comment (vankop) [reply packages/grafana-ui/.storybook/webpack.config.js:10]: yes good idea.

**Comment by aknuds1** [comment]
:crossed_fingers: this work can be merged, since StoryBook is really slowing down CI at the moment :)

**Comment by vankop** [comment]
I am busy right now, feel free to pick it up + will be great to see incremental build improvments

**Comment by aknuds1** [comment]
Unfortunately I'm not involved in the front-end development, so I don't know how to tweak the build :/

**Comment by zoltanbedi** [comment]
Thanks for contributing @vankop. Let me share my 2c here. I think we should upgrade storybook to v6 first and then revisit the build performance. I know that the main problem is the props table generator because you disabled it for development it means that the ci is still going to be slow. All in all I like the idea to make this faster but I think we should upgrade to v6 first.

**Comment by aknuds1** [comment]
What @zoltanbedi says sounds sensible.

**Comment by mckn** [comment]
A quick update: @peterholmberg is currently looking into upgrading storybook 6 and we will see what bits and pieces from this PR we can apply after that is done.

I love this initiative and great that we could speed up things some more.

**Comment by mckn** [comment]
As I mentioned earlier we are looking into upgrading Storybook to version 6 so we will close this PR. We would love to get some help of speeding up the upgraded version of Storybook once that is in place.

Issue: https://github.com/grafana/grafana/issues/25312

Thanks again for opening this PR 🙌
